src/Raven.Server/Documents/DocumentTombstoneCleaner.cs

```diff
@@ -0,0 +1,103 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+
+using Raven.Abstractions.Logging;
+
+using Sparrow.Collections;
+
+namespace Raven.Server.Documents
+{
+    public class DocumentTombstoneCleaner : IDisposable
+    {
+        private static readonly ILog Log = LogManager.GetLogger(typeof(DocumentTombstoneCleaner));
+
+        private readonly object _locker = new object();
+
+        private bool _disposed;
+
+        private readonly DocumentDatabase _documentDatabase;
+
+        private readonly ConcurrentSet<IDocumentTombstoneAware> subscriptions = new ConcurrentSet<IDocumentTombstoneAware>();
+
+        private Timer _timer;
+
+        public DocumentTombstoneCleaner(DocumentDatabase documentDatabase)
+        {
+            _documentDatabase = documentDatabase;
+        }
+
+        public void Initialize()
+        {
+            _timer = new Timer(ExecuteCleanup, null, TimeSpan.FromMinutes(1), _documentDatabase.Configuration.Tombstones.Interval.AsTimeSpan);
+        }
+
+        public void Subscribe(IDocumentTombstoneAware subscription)
+        {
+            subscriptions.Add(subscription);
+        }
+
+        public void Unsubscribe(IDocumentTombstoneAware subscription)
+        {
+            subscriptions.TryRemove(subscription);
+        }
+
+        internal void ExecuteCleanup(object state)
+        {
+            if (Monitor.TryEnter(_locker) == false)
+                return;
+
+            try
+            {
+                var tombstones = subscriptions
+                .SelectMany(x => x.GetLastProcessedDocumentTombstonesPerCollection())
```

Comment:
What about tombstones where we don't have anyone watching? They'll not be deleted.